INDI/INDI/indiserver.c

```diff
@@ -1948,150 +1958,211 @@ static int readFromDriver(DvrInfo *dp)
+        {
+            if (sendClientMsg(cp) < 0) { return -1; }/* fds effected */
+            s--;
         }
     }
+    return s;
+}
 
+static int handle_driver_messages(int s, fd_set* prs, fd_set* pws)
+{
+    int i;
     /* message to/from driver? */
     for (i = 0; s > 0 && i < ndvrinfo; i++)
     {
-        DvrInfo *dp = &dvrinfo[i];
-        if (dp->active)
+        DvrInfo *dp = dvrinfo + i;
+        if (!dp->active || dp->restartDelayus) { continue; }
+
+        if (s > 0 && FD_ISSET(dp->rfd, prs))
         {
-#           if 0
-            if (dp->pid != REMOTEDVR && FD_ISSET(dp->efd, &rs))
-            {
-                if (stderrFromDriver(dp) < 0)
-                    return; /* fds effected */
-                s--;
-            }
-#           endif/*0*/
-            if (s > 0 && FD_ISSET(dp->rfd, &rs))
-            {
-                if (readFromDriver(dp) < 0)
-                    return; /* fds effected */
-                s--;
-            }
-            if (s > 0 && FD_ISSET(dp->wfd, &ws) && nFQ(dp->msgq) > 0)
-            {
-                if (sendDriverMsg(dp) < 0)
-                    return; /* fds effected */
-                s--;
-            }
+            if (readFromDriver(dp) < 0) { return -1; }/* fds effected */
+            s--;
+        }
+        if (s > 0 && FD_ISSET(dp->wfd, pws) && nFQ(dp->msgq) > 0)
+        {
+            if (sendDriverMsg(dp) < 0) { return -1; }/* fds effected */
+            s--;
         }
     }
 }
 
+/* Service traffic from cmd FIFO, client socket, clients, and drivers */
+static void indiRun(void)
+{
+    fd_set rs, ws;
+    int maxfd = 0;
+    int s;
+    struct timeval tv;
+
+    /* Init read and write FD sets with no writers or readers */
+    FD_ZERO(&ws);
+    FD_ZERO(&rs);
+
+    /* Set up active I/O channels */
+    maxfd = setup_command_FIFO(maxfd, &rs);
+    maxfd = setup_new_client_connect(maxfd, &rs);
+    maxfd = setup_client_messages(maxfd, &rs, &ws);
+    maxfd = setup_driver_messages(maxfd, &rs, &ws);
+
+    tv.tv_sec = SELECT_WAITs;
+    tv.tv_usec = 0;
+
+    /* Wait for I/O action; s is count of I/O channels ready for data*/
+    errno = 0;
+    s = select(maxfd + 1, &rs, &ws, NULL, &tv);
+    //fprintf(stderr, "%s: select(...) = %d\n", indi_tstamp(NULL), s);
+    if (s < 0)
+    {
+        if(errno == EINTR) { return; }
+        fprintf(stderr, "%s: select(%d): %s\n", indi_tstamp(NULL), maxfd + 1, strerror(errno));
+        Bye();
+    }
+
+    /* Call handlers, which act on select(...) results */
+    /* Return if these handlers return a negative count of channels */
+    if ((s = handle_command_FIFO(s, &rs)) < 0) { return; }
+    if ((s = handle_new_client_connect(s, &rs)) < 0) { return; }
+    if ((s = handle_client_messages(s, &rs, &ws)) < 0) { return; }
+    /* Do not return for driver handlers returning negative count ... */
+    s = handle_driver_messages(s, &rs,& ws);
+
+    /* ... so the restart list will be processed */
+    handle_restart_list(&tv);
```

Comment:
The routine call, handle_restart_list, is what implements the 10s restart delay.